src/time.rs

```diff
@@ -0,0 +1,189 @@
+/// since an arbitrary moment in time, such as system startup.
+///
+/// * A value of `0` is inherently arbitrary.
+/// * A value less than `0` indicates a time before the starting
+///   point.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Instant {
+    millis: i64,
+}
+
+impl Instant {
+    /// Create a new `Instant` from a number of milliseconds.
+    pub fn from_millis(millis: i64) -> Instant {
+        Instant { millis }
+    }
+
+    /// The fractional number of milliseconds that have passed
+    /// since the beginning of time.
+    pub fn millis(&self) -> i64 {
+        self.millis % MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds that have passed since the
+    /// beginning of time.
+    pub fn secs(&self) -> i64 {
+        self.millis / MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds that have passed since
+    /// the biginning of time.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl fmt::Display for Instant {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis + rhs.total_millis())
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis - rhs.total_millis())
+    }
+}
+
+/// A relative amount of time.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    millis: i64,
+}
+
+impl Duration {
+    /// Create a new `Duration` from a number of milliseconds.
+    pub fn from_millis(millis: i64) -> Duration {
+        Duration { millis }
+    }
+
+    /// The fractional number of milliseconds in this `Duration`.
+    pub fn millis(&self) -> i64 {
+        self.millis % MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds in this `Duration`.
+    pub fn secs(&self) -> i64 {
+        self.millis / MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds in this `Duration`.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl ops::Add<Duration> for Duration {
+    type Output = Duration;
+
+    fn add(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis + rhs.total_millis())
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Duration;
+
+    fn sub(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis - rhs.total_millis())
+    }
+}
+
+impl ops::Mul<Duration> for Duration {
+    type Output = Duration;
+
+    fn mul(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis * rhs.total_millis())
+    }
+}
+
+impl ops::Div<Duration> for Duration {
+    type Output = Duration;
+
+    fn div(self, rhs: Duration) -> Duration {
```

Comment:
You're making the same mistake here as with `Mul`, but it's more subtle. In principle, `Duration / Duration` is a meaningful operation (unlike `Duration * Duration`, which returns a physically nonsensical quantity, this returns the ratio of the two durations), but it's not *especially* useful. Instead, this should be an `impl ops::Div<u32> for Duration`.

Reply:
Same as above.